File: packages/shared-contracts/src/dto.ts
/**
 * M0 数据传输对象(Zod schema + 推导类型)。
 * 红线#4:凭据相关响应 schema 永不含明文 token —— CredentialInfoSchema 只暴露
 * configured / fingerprint;PUT 请求体的 token 仅入参,绝不出现在任何响应/SSE schema。
 */
import { z } from 'zod';
import { CAPABILITIES } from './capabilities.js';
import {
  PROVIDERS,
  PROVIDER_STATUSES,
  DATASETS,
  JOB_TYPES,
  JOB_STATUSES,
  JOB_TRIGGERS,
  BOARDS,
  ONBOARDING_STEPS,
} from './enums.js';

const ProviderEnum = z.enum(PROVIDERS);
const CapabilityEnum = z.enum(CAPABILITIES);

/** 能力声明 schema(对前端/DB 的 caps_json 形态)。类型 CapsRecord 由 capabilities.ts 统一导出。 */
export const CapsRecordSchema = z.record(CapabilityEnum, z.boolean());

// ── 健康 ────────────────────────────────────────────────────────────────
export const HealthSchema = z.object({
  status: z.enum(['ok', 'degraded', 'down']),
  db_ok: z.boolean(),
  engine_ok: z.boolean(),
  gpu: z.boolean().optional(),
  version: z.string().optional(),
});
export type Health = z.infer<typeof HealthSchema>;

// ── 引导 ────────────────────────────────────────────────────────────────
export const OnboardingStateSchema = z.object({
  done: z.boolean(),
  step: z.enum(ONBOARDING_STEPS),
  active_provider: ProviderEnum.optional(),
});
export type OnboardingState = z.infer<typeof OnboardingStateSchema>;

// ── 数据源 ──────────────────────────────────────────────────────────────
export const RateLimitSchema = z.object({
  per_min: z.number().int().positive(),
  concurrency: z.number().int().positive().optional(),
});
export type RateLimit = z.infer<typeof RateLimitSchema>;

export const ProviderInfoSchema = z.object({
  id: ProviderEnum,
  display_name: z.string(),
  caps: CapsRecordSchema,
  needs_token: z.boolean(),
  rate_limit: RateLimitSchema.optional(),
  priority: z.number().int(),
  status: z.enum(PROVIDER_STATUSES),
  last_check_at: z.string().nullable().optional(),
});
export type ProviderInfo = z.infer<typeof ProviderInfoSchema>;

/** PUT /providers/:id/credential 请求体。token 仅入参 —— 立即加密,绝不回显。 */
export const CredentialPutReqSchema = z.object({
  token: z.string().min(1),
});
export type CredentialPutReq = z.infer<typeof CredentialPutReqSchema>;

/** 凭据信息响应 —— 红线:永不含明文 token。 */
export const CredentialInfoSchema = z.object({
  configured: z.boolean(),
  /** 明文 token 的 SHA-256 前 8 位,仅供 UI 显示「已配置 ****a1b2」,不可逆。 */
  fingerprint: z.string().nullable(),
});
export type CredentialInfo = z.infer<typeof CredentialInfoSchema>;

/** POST /providers/:id/test 结果。 */
export const ProviderTestResultSchema = z.object({
  status: z.enum(PROVIDER_STATUSES),
  latency_ms: z.number().optional(),
  caps: CapsRecordSchema,
  rate_limit: RateLimitSchema.optional(),
  /** Tushare 积分提示(探测得到的近似值)。 */
  points_hint: z.number().nullable().optional(),
  /** 降级说明(非空即提示用户预期更低)。 */
  degraded: z.array(z.string()),
  message: z.string().optional(),
});
export type ProviderTestResult = z.infer<typeof ProviderTestResultSchema>;

// ── 作业 / 建缓存 ───────────────────────────────────────────────────────
export const CacheBuildParamsSchema = z.object({
  universe: z.object({
    boards: z.array(z.enum(BOARDS)).min(1),
    liq_min_amount_yi: z.number().nonnegative().optional(),
    codes: z.array(z.string()).optional(),
  }),
  years: z.number().int().positive().optional(),
  start_year: z.number().int().optional(),
  datasets: z.array(z.enum(DATASETS)).min(1),
  rate: RateLimitSchema.optional(),
  resume: z.boolean().optional(),
});
export type CacheBuildParams = z.infer<typeof CacheBuildParamsSchema>;

export const JobCreateReqSchema = z.object({
  type: z.enum(JOB_TYPES),
  trigger: z.enum(JOB_TRIGGERS).optional(),
  provider: ProviderEnum.optional(),
  params: z.record(z.string(), z.unknown()).optional(),
});
export type JobCreateReq = z.infer<typeof JobCreateReqSchema>;

export const JobPatchReqSchema = z.object({
  action: z.enum(['pause', 'resume', 'cancel']),
});
export type JobPatchReq = z.infer<typeof JobPatchReqSchema>;

export const JobInfoSchema = z.object({
  id: z.string(),
  type: z.enum(JOB_TYPES),
  provider: ProviderEnum.nullable().optional(),
  status: z.enum(JOB_STATUSES),
  trigger: z.enum(JOB_TRIGGERS),
  total: z.number().int().nonnegative(),
  done_count: z.number().int().nonnegative(),
  failed_count: z.number().int().nonnegative(),
  progress: z.number().min(0).max(1),
  cursor: z.record(z.string(), z.unknown()).nullable().optional(),
  rate_limit: z.record(z.string(), z.unknown()).nullable().optional(),
  error: z.string().nullable().optional(),
  started_at: z.string().nullable().optional(),
  finished_at: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
});
export type JobInfo = z.infer<typeof JobInfoSchema>;

/** SSE 进度事件(GET /events/jobs/:id)。 */
export const JobProgressEventSchema = z.object({
  job_id: z.string(),
  status: z.enum(JOB_STATUSES),
  progress: z.number().min(0).max(1),
  total: z.number().int().nonnegative(),
  done_count: z.number().int().nonnegative(),
  failed_count: z.number().int().nonnegative().optional(),
  stage: z.string().optional(),
  message: z.string().optional(),
  /** 限流冷却剩余秒数,UI 显示「限流中,预计 x 秒后继续」。 */
  rate_cooldown_s: z.number().nonnegative().optional(),
  ts: z.string(),
});
export type JobProgressEvent = z.infer<typeof JobProgressEventSchema>;

// ── 缓存覆盖台账 ────────────────────────────────────────────────────────
export const CoverageRowSchema = z.object({
  stock_code: z.string(),
  dataset: z.enum(DATASETS),
  provider: ProviderEnum,
  first_date: z.string().nullable(),
  last_date: z.string().nullable(),
  rows: z.number().int().nonnegative(),
});
export type CoverageRow = z.infer<typeof CoverageRowSchema>;

export const CoverageSummarySchema = z.object({
  stock_count: z.number().int().nonnegative(),
  first_date: z.string().nullable(),
  last_date: z.string().nullable(),
  total_rows: z.number().int().nonnegative(),
  bytes: z.number().int().nonnegative().optional(),
  by_dataset: z.array(CoverageRowSchema).optional(),
});
export type CoverageSummary = z.infer<typeof CoverageSummarySchema>;

// ── 统一错误体 ──────────────────────────────────────────────────────────
export const ApiErrorSchema = z.object({
  error: z.object({
    code: z.string(),
    message: z.string(),
    detail: z.unknown().optional(),
    job_id: z.string().optional(),
  }),
});
export type ApiError = z.infer<typeof ApiErrorSchema>;

File: packages/shared-contracts/python/sinan_contracts/dto.py
"""M0 DTO Python 绑定(pydantic v2)—— 与 TS Zod schema 对齐。

红线#4:凭据响应模型 CredentialInfo 永不含明文 token,只暴露 configured/fingerprint。
"""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

from .enums import (
    Board,
    Dataset,
    JobStatus,
    JobTrigger,
    JobType,
    OnboardingStep,
    Provider,
    ProviderStatus,
)

CapsRecord = dict[str, bool]


class Health(BaseModel):
    status: Literal["ok", "degraded", "down"]
    db_ok: bool
    engine_ok: bool
    gpu: Optional[bool] = None
    version: Optional[str] = None


class OnboardingStateDTO(BaseModel):
    done: bool
    step: OnboardingStep
    active_provider: Optional[Provider] = None


class RateLimit(BaseModel):
    per_min: int = Field(gt=0)
    concurrency: Optional[int] = Field(default=None, gt=0)


class ProviderInfo(BaseModel):
    id: Provider
    display_name: str
    caps: CapsRecord
    needs_token: bool
    rate_limit: Optional[RateLimit] = None
    priority: int
    status: ProviderStatus
    last_check_at: Optional[str] = None


class CredentialPutReq(BaseModel):
    token: str = Field(min_length=1)


class CredentialInfo(BaseModel):
    """红线:永不含明文 token。"""

    configured: bool
    fingerprint: Optional[str] = None


class ProviderTestResult(BaseModel):
    status: ProviderStatus
    latency_ms: Optional[float] = None
    caps: CapsRecord
    rate_limit: Optional[RateLimit] = None
    points_hint: Optional[int] = None
    degraded: list[str] = Field(default_factory=list)
    message: Optional[str] = None


class CacheBuildUniverse(BaseModel):
    boards: list[Board] = Field(min_length=1)
    liq_min_amount_yi: Optional[float] = Field(default=None, ge=0)
    codes: Optional[list[str]] = None


class CacheBuildParams(BaseModel):
    universe: CacheBuildUniverse
    years: Optional[int] = Field(default=None, gt=0)
    start_year: Optional[int] = None
    datasets: list[Dataset] = Field(min_length=1)
    rate: Optional[RateLimit] = None
    resume: bool = True


class JobCreateReq(BaseModel):
    type: JobType
    trigger: Optional[JobTrigger] = None
    provider: Optional[Provider] = None
    params: Optional[dict[str, Any]] = None


class JobPatchReq(BaseModel):
    action: Literal["pause", "resume", "cancel"]


class JobInfo(BaseModel):
    id: str
    type: JobType
    provider: Optional[Provider] = None
    status: JobStatus
    trigger: JobTrigger
    total: int = 0
    done_count: int = 0
    failed_count: int = 0
    progress: float = 0.0
    cursor: Optional[dict[str, Any]] = None
    rate_limit: Optional[dict[str, Any]] = None
    error: Optional[str] = None
    started_at: Optional[str] = None
    finished_at: Optional[str] = None
    created_at: str
    updated_at: str


class JobProgressEvent(BaseModel):
    job_id: str
    status: JobStatus
    progress: float = Field(ge=0, le=1)
    total: int = 0
    done_count: int = 0
    failed_count: Optional[int] = None
    stage: Optional[str] = None
    message: Optional[str] = None
    rate_cooldown_s: Optional[float] = None
    ts: str


class CoverageRow(BaseModel):
    stock_code: str
    dataset: Dataset
    provider: Provider
    first_date: Optional[str] = None
    last_date: Optional[str] = None
    rows: int = 0


class CoverageSummary(BaseModel):
    stock_count: int = 0
    first_date: Optional[str] = None
    last_date: Optional[str] = None
    total_rows: int = 0
    bytes: Optional[int] = None
    by_dataset: Optional[list[CoverageRow]] = None
